pkg/installer/migrate.go

```diff
@@ -0,0 +1,356 @@
+
+		for key, value := range result.Drive.Labels {
+			switch key {
+			case legacyAccessTierLabelKey:
+			case legacyCreatedByLabelKey:
+			case legacyNodeLabelKey:
+			case legacyPathLabelKey:
+			case legacyVersionLabelKey:
+			default:
+				if strings.HasPrefix(key, legacyclient.GroupName) {
+					key = strings.Replace(key, legacyclient.GroupName, consts.GroupName, 1)
+				}
+				drive.Labels[key] = value
+			}
+		}
+
+		for _, finalizer := range result.Drive.Finalizers {
+			if strings.HasPrefix(finalizer, legacyclient.GroupName) {
+				finalizer = strings.Replace(finalizer, legacyclient.GroupName, consts.GroupName, 1)
+			}
+
+			if !utils.Contains(drive.Finalizers, finalizer) {
+				drive.Finalizers = append(drive.Finalizers, finalizer)
+			}
+		}
+
+		existingDrive, err := client.DriveClient().Get(ctx, string(driveID), metav1.GetOptions{})
+		if err != nil {
+			switch {
+			case apierrors.IsNotFound(err):
+				if !dryRun {
+					_, err = client.DriveClient().Create(ctx, drive, metav1.CreateOptions{})
+					if err != nil {
+						legacyDriveErrors[result.Drive.Name] = fmt.Errorf(
+							"unable to create drive %v by migrating legacy drive %v; %w",
+							driveID, result.Drive.Name, err,
+						)
+					}
+				}
+			default:
+				driveErrors[string(driveID)] = fmt.Errorf(
+					"unable to get drive by drive ID %v; %w", driveID, err,
+				)
+				delete(driveMap, result.Drive.Name)
+			}
+		} else {
+			if existingDrive.IsMigrated() {
+				legacyDriveErrors[result.Drive.Name] = fmt.Errorf(
+					"legacy drive %v is already migrated to drive %v",
+					result.Drive.Name, existingDrive.Name,
+				)
+			} else {
+				legacyDriveErrors[result.Drive.Name] = fmt.Errorf(
+					"unable to migrate legacy drive %v; drive %v already exists",
+					result.Drive.Name, existingDrive.Name,
+				)
+			}
+		}
+	}
+
+	return driveMap, legacyDriveErrors, driveErrors, nil
+}
+
+func migrateVolumes(ctx context.Context, driveMap map[string]string, dryRun bool) (legacyVolumeErrors map[string]error, volumeErrors map[string]error, err error) {
+	ctx, cancelFunc := context.WithCancel(ctx)
+	defer cancelFunc()
+
+	legacyVolumeErrors = map[string]error{}
+	volumeErrors = map[string]error{}
+
+	for result := range legacyclient.ListVolumes(ctx) {
+		if result.Err != nil {
+			return legacyVolumeErrors, volumeErrors, fmt.Errorf(
+				"unable to get legacy volumes; %v", result.Err,
+			)
+		}
+
+		fsuuid, found := driveMap[result.Volume.Status.Drive]
+		if !found {
+			legacyVolumeErrors[result.Volume.Name] = fmt.Errorf(
+				"referring drive %v of volume %v not found",
+				result.Volume.Status.Drive, result.Volume.Name,
+			)
+			continue
+		}
+
+		name := result.Volume.Name
+		nodeID := directpvtypes.NodeID(result.Volume.Status.NodeName)
+		driveID := directpvtypes.DriveID(fsuuid)
+		driveName := directpvtypes.DriveName(result.Volume.Labels["direct.csi.min.io/drive-path"])
+		size := result.Volume.Status.TotalCapacity
+
+		volume := types.NewVolume(
```

Comment:
the migrated volume will always be in pending state here. We can make it to Ready state if the stagingtargetpath and containerpath isn't empty.